### DISCUSSION TITLE
Feature Request - Relay Server/client

### DISCUSSION THREAD

**Discussion by AndyMcProducer** [jamulussoftware/jamulus#1031] [Ideas]
Hiya I can't see where to put a feature request, Corrados could you move this please  to the right department. :)

Would it be possible to have an option to set jamulus client or server as a relay server.
So say ya got a server in the USA and one in Europe, the relay would connect to a server then relay the audio and chat to and from another server. Only a single stream would be needed, we could still communicate for remote to lower or up their volumes.
I see the server records so gathered there is an audio stream there to be had for relaying.
They'd show as a connect in the server as say (Remote Server Us Oldies -USA) but chat could show as that person and their name somit like Us Oldies - Jackie: Hiya All.

Can you see the possibilities of this and advantages?

With it could be allowing so many slots from total for relays.

This could be done manually but added ms and complex setup for many to route a jamulus to a jamulus.
Would help also to make bigger servers with split resources and with the group, mute and solo options would be
pretty easy to handle.
Also doing it via jamulus clients wouldn't carry over the chat.

A relay server could be run also by someone who has much better pings, meaning a much shorter route for the audio combined. It would probs make more sense having it in the server, but then maybe client to so if someone doing the relay server could then adjust volumes or even int he future having a limiter on each connected in the relay server/client.

Andy.

**Comment by ann0see** [comment]
Hi @AndyMcProducer! Thanks for your feature request!

I think there already was a discussion on this topic (server as relay as alternative to the multi threading approach) but it wasn't implemented since there might be too much latency and syncronisation issues. Not being able to control every participant on your own mix isn't useful and if you want to control everybody, you could just use a powerful server somewhere where the ping time is almost equal to both continents. The client approach sounds interesting and might somehow be possible via a headless client on a server which then connects to another server. 

I think the problem you want to solve with this idea is that it's not easy to route audio from server A to server B?

**Comment by hoffie** [comment]
I'm wondering if this could just be a special, extra-fast, network-only Jamulus client which does not do any audio processing and just forwards the data between two servers. It would even be possible to relay chat messages (although it would still be noticable that they are forwarded). Such a special client could either be a fork of Jamulus, make use of Jamulus' protocol parsing as a library or it could even be implemented in any other language.

@AndyMcProducer What specific use cases do you have in mind?
Having larger groups (70-80 people?) seem to be no problem nowadays (I have personal experience with up to 25 people). If latency between participants is too large, a relay will probably not help. So I suspect that you are after something else?

**Reply by AndyMcProducer** [reply]
Live Jams and Friends over the pond situations. So audio could be muted by others if needed of the relay server either side, they'd say all be low ping their end, us out end, then both ends only have 1 combined stream to follow and not multi pinged connections. 
Pings are different for so many dependent on isp and hops, going over the pond this is even more wider. Using a Relay server keeps things much tighter as only a overall audio mix of all needs sending, slightly delayed wouldn't be an issue as the side leading would mute the audio from the relay and the non leading side would play to what they hear then this side can always be pushed live making all better quality and more sync'd sounding on the end result. Possible one day mics could be separated from music and on a separate channel from the relay so they could be left active when delayed leading side relay is muted. 

So sync'ign in time is possible, it's just not possible to follow them then them get the audio back in time on time. 
So which ever end starts the Jam and leads it, they can't hear back the other end but on the live stream you got it all sounding sync'd. Much simpler to handle one feed from a long distance server than multiple connections all 80 to 250 pings which is the case from multiple isp and connections from usa. Stuff like using opus multiple channel options which send as 1 packet would be best, then multiple can be received in time the other end but as 1 like user input.

**Reply by AndyMcProducer** [reply]
BTW you could allow for more relays, same applies, leading server can't hear others but end server would hear all in time. So if USA was relayed to Poland and UK was relayed to USA, The audio route would be poland to usa, usa to uk, or uk to usa, usa to uk. Always though the leading server can't have audio back from the others, only the end server can have all audio but in sync to. Unless pings were low enough or peeps can play with long pings.

**Reply by ghost** [reply]
What @AndyMcProducer says in the two previous comments is complete, reasonable, and understandable. It is possible to do the equivalent with two internet connected computers, each running a Jamulus client and connected with an audio patch cable as such:
- some musicians in a Europeean city connect to a Jamulus server in that city.
- some musicians in a North American city connect to a Jamulus server in their city.
- somebody in Eurpoe connects to the Europe session with one computer, and also connects to the other session with the other computer.
this somebody connects the headphones output on the European computer to the microphone input of the other computer.
Now the North Americans can hear the Europeans and play along in a time delayed fashion.

**Comment by gilgongo** [comment]
@AndyMcProducer  Hi - just moving this to a discussion until we can firm it up into a defined backlog item we can implement.

**Comment by gene96817** [comment]
@AndyMcProducer    I am not understanding clearly the motivation for your proposed solution. Could you discuss the following:
1- scenario and problem that the relay network would solve. What is the before and after behavior?
     I'd like to work through the problem and benefit for different participants at different parts of the relay'ed network. I'd like to be convinced the benefit you propose applies to all participants at any part of the realy'ed network. I am wondering if you are just moving the problem around.
2- who sets up the relay network? is this a mode where these are public servers that are configured for a relay'ed session? Is this a network where a single owner has to set up each and every server in the relay configuration?
3- What control is needed over a distant network in the relay network? 
4- If the servers in the relay network are owned by different individuals, what access do they need to give to the person conducting a relay'ed session?

**Reply by AndyMcProducer** [reply]
I guess to keep it simple as possible any server could initiate or receive a relay, a pw both ends of the relay to allow connection.
End server naturally would receive all servers audio in sync, this would be better for streaming live.
But with servers in the relay'd connection could take turns in muting others relayed audio allowing for jams and chat to still be as one but at the cost of some return audio being cut to others in the relayed chain. So if 2 relayed servers close then say one on an 80 or more ping, the 80 ping server could jam along but their audio would be muted to the other 2, or the 80 ping server can lead then the other 2 would be muted to them. Maybe a lead option could be added so it auto muted audio from others that are outside of a ping scope. 
It could be achieved manually but not for all and not with out a lot more setup and chance of failure.
Although chat couldn't with out again some sort of curling of the chat re-pushed into the main chat.

Another fix this does is people could even be relayed multi channel as opus supports this to 1 packet for all. So mic and instrument channels could be all sent to the other server as one packet one connection, each relay would consist of local ping rates, chat would still be connected as one, all would be seen but some would be relayed in.
So on server 1, say fred, joe and julie are all on a sub 30 ping, now freds ping to server 2 is 150, julies is 105 and joe has 80, for them to all connect to the other server would mean not much chance of them jamming so they couldn't lead as a group of 3.
Now lets say all 3 are connected to server 1 sub 30 ping and server one relayed to server 2 is between 80 and 90 ping. Now server one can play and server 2 has them all sync'd and can now play along, but as server 1 is the lead they'd have to mute server 2 instrument channels. But this also allows for multiple bass players, drummers guitarist, as parts of the relay will need to mute each other. So in server 2 harry is singing, they are lead and there are durms bass guitar, they have server 1 instruments muted (instruments includes singing, mic is for chat), but server 1 can hear server 2, they now play in sync to server 2 and julie can sing alongside harry from server 2, even though harry can't hear her but if server 1 now streams the audio to the net live then julie and harry sound like they are singing at the same time no lag. Because the relay is just that, a relay, it can be higher ms and bitrate as delay isn't as important as one relay or multiple just follow the lead .

So best solution is if a server 1 is relayed to server 2, a 3rd server connects to server 1, then server 3 shoudl also relay to server 2.
Reason being is if a lead option was created then who ever is lead other 2 servers become relay 2 and 3, so one server gets all servers audio. You could keep it 1 to 2 to 3 but then if 2 leads they supply 1 and 3 but 1 and 3 don't hear each other. So by circling the relay if 2 leads, the audio would then go to say 2 then 1 meaning last server gets all 3, 2nd server gets 2.
I think probably if it was just 1 to 2 to 3 then this would be fine to but no one server then would have the complete audio to stream.

Some might thing what's the point, but musicians and singers want to perform their craft's. Some may want to just jam but others will want to perform, but even just jamming this lets friends still connect for a jam which can be recorded/saved across the world. 

No one is going to make something anytime soon that can go faster than light, so until they can send direct through the earth which is being created, then, these alternatives need to be considered.

Add a spin to it even more, the relays could actually be like ninjam and fake time sync'd then with a big enough interval all could heard each other and in sync but be locally connected in realtime.

**Reply by gene96817** [reply]
I have trouble with "fake time sync". I personally don't see how faking sync helps.
I am also having trouble understanding why it makes sense to jam together when you can't hear some of the other participants.

One way to make the sync problem explicit is to create a timing ladder diagram (to borrow a tool from network protocol design).

> So in server 2 harry is singing, they are lead and there are durms bass guitar, they have server 1 instruments muted (instruments includes singing, mic is for chat), but server 1 can hear server 2, they now play in sync to server 2 and julie can sing alongside harry from server 2, even though harry can't hear her but if server 1 now streams the audio to the net live then julie and harry sound like they are singing at the same time no lag.

This sounds like a variation on overdubbing a recording. Harry can never perform interactively with Julie. Julie is just overlaying Harry's sound. If this is a goal, then you would describe the relay chain of servers so that musicians at the root server play on autopilot relative to musicians down the relay chain. Each node in the chain overlay sound as it moved down the chain. Yes at each step you can overlay sound that is synced. To me, this isn't jamming.

**Comment by dcorson-ticino-com** [comment]
I fail to see how relay servers could help any of the problematic use cases we might have.
Network bandwidth is not a major problem in this day and age.
The program allows jamming with over 100 participants if one (100) wishes.  I have personally used up to 24 and it worked fine.
If a relay server could add time warp so that the packets arrive at the same time as they are sent it would be a dream.
Unfortunately we are bound by the laws of physics.
Adding any processing by additional servers along the route from one client to the next will only add latency.
This is a good idea to think through, but once thought threw, I think it is an idea to reject.

**Comment by AndyMcProducer** [comment]
People said that about the wheel once to, and that the Earth was flat.

Multiple connects will hit your ping eventually and have an impact if that's what ya was implying, more so on some isp's that throttle when many connections are made.

Some replies here show you didn't read it all and only skipped it, yes is a lot but best to read it if your gonna say how wrong it is.

Fake time would only be handy on the relay for some stuff, not fake time on the servers themselves.

I explained the physics of ping being speed of light and would not change.

IF you want to jam with high pings then as said physics prevents this. People say i can jam on high ping, great for you but bad for all else who have to hear you way out of sync and mute you.
But in a relay least one side hears the whole jam and as it's multiple servers then normally would be enough to fill each end meanign the server muted woudl still have plenty of jam.
But for live and multiple singers performing they could achieve this much more in sync.

I don't think an option shouldn't be implemented because some don't get it and others can't see why it would be needed.
I'm thinking this for ALL not the odd winger.

AND no processing is being adding from client to client, actually less ms from relay to relay would be achieved for all as an average.
So bill, ben and a flower pot man on server one with 30 ping, jack and jill wih a guy named phil on the server two with a 30ish ping.
Bill Ben and the flower pot man connect to server two directly, Bill has a ping 155, Ben has a ping of 95 and the flower pot man has a ping of 250.
BUT now Bill Ben and the flower pot man reconnect to server one with their nice 30 pings, now server one connects via relay to server two and has a ping of 80.
Now Bill and Ben and the Flower Pot Man are all in sync and their combined audio is sent in sync on a ping of 80 to server two. Server One is half a choir and server two the other half.
They are about to put on a show, server one starts a count in then start singing, server 2 now has all 3 in sync and sing along.
Now Server two streams it live on the internet to their audience, all in time, but server one will have to mute server two so they don't head them back delayed.
The show must go on.
Side 2 could even prerecord themselves and which could be a local server playback track, so a pre record of server 2's parts could be played for server 1 to sing to, if they needed it.
________________________________

**Reply by gene96817** [reply]
> People said that about the wheel once too, and that the Earth was flat.

> Some replies here show you didn't read it all and only skipped it, yes is a lot but best to read it if your gonna say how wrong it is.

Please, we are all trying to be helpful. Our style here is terse and with a tricky idea, it would be really helpful to lay out the idea in the long, step-by-step explanation. Otherwise, a dummy like me might miss how clever your idea really is.  I really don't want to read your proposal as an engineering problem set to reverse engineer the missing details.

> IF you want to jam with high pings then as said physics prevents this. People say i can jam on high ping, great for you but bad for all else who have to hear you way out of sync and mute you.

I don't understand why physics prevents jamming with high ping times. It is easy to add an arbitrary amount of delay to satisfy anyone that want high ping times. Physics does establish the floor on minimum ping times.

> But in a relay least one side hears the whole jam and as it's multiple servers then normally would be enough to fill each end meanign the server muted woudl still have plenty of jam.
But for live and multiple singers performing they could achieve this much more in sync.

I would really like to understand the scenario (or scope) where you have solved this problem.

> So bill, ben and a flower pot man on server one with 30 ping, jack and jill wih a guy named phil on the server two with a 30ish ping.
Bill Ben and the flower pot man connect to server two directly, Bill has a ping 155, Ben has a ping of 95 and the flower pot man has a ping of 250.
BUT now Bill Ben and the flower pot man reconnect to server one with their nice 30 pings, now server one connects via relay to server two and has a ping of 80.

Thank you for some details.
What I see here is:
Ping (server1, server 2) =80ms
Ping (Bill, server 2) = 155ms
Ping (Ben, server 2) = 95 ms
Ping (FlowerPot, server 2) = 250 ms
Ping (Bill, server1) = Ping (Ben, server 1) = Ping (FlowerPot, server 1) = 30 ms

Approximating all distance as linear and applying Phytagorean theorem,
Ping (Bill, server 1)^2 + Ping (server 1, server 2)^2 = Ping (Bill, server2)^2
30ms^2 + 80ms^2 /= 155ms^2
900+6400-24025=-16725
you magically erased 129ms with relay for Bill.
similar math should/can be done for Ben and Mr. FlowerPot. 
There is a timewarp in this architecture OR my rough model is very wrong.

> Now Server two streams it live on the internet to their audience, all in time, but server one will have to mute server two so they don't head them back delayed.

At this point musicians at server 1 is not performing with musicians at server2.  This really isn't jamming. Might as well do a virtual performance with recording magic.

**Comment by AndyMcProducer** [comment]
Yep, even fire once was thought of as magic. Pings are mostly dictated by hops, so yes you magically cut the latency to that of the server connection which most run on aws, google and so on.
Hops from within their own network tends to be even better.

I guess if it's implemented, you'll be convincing everyone is MAGIC! Except those who understand how it works.
:S
________________________________

**Reply by gene96817** [reply]
@AndyMcProducer   You are not taking me seriously. Even if the routers were relatively slow and added 5ms. If you were eliminating 5 routers or 25 ms, you will have to explain how you are erasing 100+ ms. That is a lot of km to erase. It's magic for people who don't understand science. I am asking you to explain how it works. Magic is not an explanation.

**Comment by AndyMcProducer** [comment]
Also, I never said those on server 1 would then get a ping of 80 to server 2, their combined audio sync'd would be on one ping of 80 to server 2.
So rather than random all over the place pings they'd all be sync'd but could still be separate channels adjustable on server 2 and vice versa.
Some can play on a higher ping, this way they only have 1 ping to deal with not 10 pings all over the place.

But you dismissed it and didn't even understand it. 😕

________________________________

**Reply by gene96817** [reply]
@AndyMcProducer   Ok... we are failing to communicate. I took the time to detail how I interpret your message. You give me the same respect and tell me where my model fails. Through that process, I would understand you.

> BUT now Bill Ben and the flower pot man reconnect to server one with their nice 30 pings, now server one connects via relay to server two and has a ping of 80.

Oh I misunderstood you. The 80 ms is the net ping time. So subtracting out 30ms (for user to server 1) , the server1-to-server2 ping time is 50. Yikes, the problem just got worst. Yes, time warp is magic (but not believable).

**Comment by AndyMcProducer** [comment]
Server 1 won't be in sync with server 2 if server one leads. But server 2 will have all users sync'd from server 1, unless people connecting to server 1 are on pings like 30, 60, 120, 157 and so on, but they wouldn't be.
You'd have people connect to the closer server to them, so you'd say have peeps within a 40 overall ping on server 1, same on server 2, so then the audio will be as sync'd as it is played but then relayed to the server 2
as 1 packet 1 connection.
Users on server 1 if they were to connect to server 2 direct could all be in the uk and have 30ish ping to server 1 but to server 2 they could be 80, 140, 190, depending on how many hops their isp's take.
So, by relaying the combined audio using multi channel support on opus you effectively have all audio sent from one server to the other, but if also both are on a cloud server, then it will be much lower ms.
This is not to negate the time it takes to travel; this is to put all in sync as they would be local with another server remote. They ain't in sync, server 2 is hearing them delayed but the audio source has all users
in sync on server 1,, again unless someone is on a stupid ping to server one, then they'd sound out on server one but sound the same amount out on server 2 too.

In laming terms, Bill, Ben and Flower Pot Man travel to a parcel and it makes them grow in size, when they all get to the parcel, they are 30cm high. In the box they go and off they travel.
It costs 80 coins to send them to America, when they get there, the Americans open the box, what size are Bill, Ben and the Flower Pot Man?


________________________________
